### ISSUE TITLE
Allow to invoke the lambda function multiple times

### ISSUE THREAD

**Issue by ibratoev** [lambci/docker-lambda#14]
In AWS Lambda containers are not destroyed after each execution.

In my scenario (tests), I need to invoke a function multiple times. It would be much faster if, you don't need to recreate the whole container, including the NodeJS process, before each invocation. 

Additionally, this can catch potential production issues as it will be closer to the way AWS Lambda works.

**Comment by mhart** [comment]
You'd need to create a whole communication layer to achieve this, instead of the existing invocation model – happy to accept a PR on it though

**Comment by ibratoev** [comment]
Closed/opened by mistake ;)
For NodeJS, the main process can listen on a Unix Socket and process events on request. It shouldn't be too hard.
I looked at the AWS code - there it is implemented with a native module, but I don't think the efficiency is that important for local usage.

**Comment by jscheid** [comment]
Hi, here is my attempt at a solution: https://github.com/jscheid/lambda-server

There's a lot to say about this... for starters, merging this into docker-lambda, rather than keeping it as a separate image, would get rid of 80% of the FIXMEs in the code. Before starting that discussion I wanted to check that there is still interest in adding this feature, and that you agree with the general thrust of the solution?

**Comment by mhart** [comment]
@jscheid have you seen AWS SAM Local? It uses docker-lambda under the hood and achieves everything you'd need (I think) https://github.com/awslabs/aws-sam-local

**Comment by jscheid** [comment]
@mhart yes I've seen it (used it actually) but no, it doesn't: https://github.com/awslabs/aws-sam-local/issues/227

As far as I know they're just using vanilla (as in, one-container-per-invocation) lambci/lambda under the hood... https://github.com/awslabs/aws-sam-local#a-special-thank-you

**Comment by jscheid** [comment]
@mhart am I missing something?

**Comment by mhart** [comment]
No, not missing anything – if you have to keep it warm as opposed to just multiple invocations, then SAM Local doesn't do this.

**Comment by jscheid** [comment]
Ah, good to know, thanks.

Are you interested at all in adding this feature to this repository, and if you don't, would you help me with a few questions I have (like [this one](https://github.com/jscheid/lambda-server/blob/0e044ca9a59966bf9908298c90efb2962a962259/server.js#L66-L68))?

**Comment by mhart** [comment]
That code comes from Lambda (you can find it on the filesystem) – which is true for all the other files on the container – the only file I add is the mock for the native library.

I think if a solution were to be added to this repo, it would need to be runtime-independent – or at least, the same interface for each runtime.

**Comment by jscheid** [comment]
Thanks for explaining this, it's what I expected but I'm still somewhat surprised. Do you happen to know how to reconcile these two observations:

1. AWS Lambda clearly performs multiple invocations per process when warm:

   - http://docs.aws.amazon.com/lambda/latest/dg/best-practices.html advocates the use of global/static variables, and of course these don't normally survive after the process exits,

   - it's hard to imagine how it could achieve such decent response times (<5 ms I believe) if it had to load a new node.js runtime, and load and parse what can potentially be hundreds or thousands of JS files in npm packages, for every invocation.

2. Yet the code in `awslambda` appears to assume only one invocation per process, for example:

   - the module executes the invocation at the top level (on load), rather than in response to some message passed in, and rather than exporting a method to be invoked by some other module,

   - the way the code deals with global event handlers and environment variables looks very much like it's relying on process exit for cleanup.

Could it be that, while awslambda/index.js can be found on in their file system, it's not actually what they use for real?

I suppose there is a chance that they do in fact use this code but with some kind of wrapper around `require()`, `process` etc., or with a special Node runtime. I can't see how they could solve it purely on the OS level though.

I'll run a few experiments on Lambda to get a better idea of how this all fits together, but I'd still be curious if you had any insights to share.

Would you agree that it would be useful if docker-lambda, in addition to what you list in the README (installed software, file structure, etc.) would also replicate this aspect accurately? It would probably allow speeding up CI tests considerably, and allow testing of any in-app caching methods.

Your point about other runtimes is well taken, but first I'd like to figure out how to solve it for just a single runtime robustly and accurately.

**Comment by mhart** [comment]
I'm not sure what makes you think their code only assumes one invocation per process? It doesn't look that way to me (or didn't last time I checked).

Certainly anything set globally will only disappear on a process exit – although a bunch of env variables are set per invocation.

The awslambda code handles pretty much everything needed for an invocation – it just relies on the native module to communicate back to the AWS Lambda supervisor (or whatever AWS calls it) – and listen for new invocations

**Comment by jscheid** [comment]
Yes, it seems like I'm missing something.

It looks to me like it's assuming one invocation per process because the code is executed at the top level and uses environment variables for arguments.

For multiple invocations in the same process, using the same Node.js runtime (for global variables etc.) what you would usually do is just model this as a JavaScript function with arguments and a return value?

And also, just to pick a random bit of code out of many similar ones, `process.on('beforeExit', () => invokeManager.finish(null, null, false));` looks a lot like the invocation is tied to the process lifetime.

Again, I might be missing something, but so far every way of invoking this code multiple inside the  same Node runtime cleanly and robustly involves all sorts of hacks, like using `eval` (or the `vm` module) to run the code, to remove all event handlers for `uncaughtException` and `beforeExit` after each invocation, etc.

**Comment by mhart** [comment]
Specifically: do a search for `awslambda.waitForInvoke` – you can see it's called at the end of each invocation (in the `finish` function) – so it just loops

**Comment by mhart** [comment]
Also I'm not sure what you mean by "executed at the top level" – each event is processed by the user's specified event handler by the `start` function (which, again, is a callback to `waitForInvoke`).

The user's code is only `require`d once (in `_getHandler`) – if that's what you mean by the top level – but that's expected – that's only supposed to be a once-off execution.

**Comment by jscheid** [comment]
Ahh... right, of course, I was thrown off by the fact that the docker image's entry point is simply invoking the code, rather than passing a message of some sort.

In this case, you wouldn't happen to know how to inject multiple invocations into `awslambda`? The supervisor communication you were referring to isn't documented anywhere?

**Comment by jscheid** [comment]
Never mind, I see now, it comes from your mock.

**Comment by jscheid** [comment]
Sorry, I didn't see the forest for the trees. Thanks for working through this with me, now I can make the code much less of a hack.

**Comment by jscheid** [comment]
One more question, are you opposed to introducing a build step, for Babel or such?

**Comment by der-flo** [comment]
@jscheid: Did you make progress?

Doing serious tests with Lambda without a fast development environment is quite impossible. Does anybody know a working solution?

**Comment by binarymist** [comment]
Seems like a very common case:
https://github.com/awslabs/aws-sam-cli/issues/227
https://github.com/awslabs/aws-sam-cli/issues/239

**Comment by devoto13** [comment]
One idea for the simple protocol is to use `stdin` for sending events and `stdout` for returning responses. The format is `\n`-separated JSON strings. This should provide the same protocol for all runtimes, however communication layer needs to be done for each runtime separately.

E.g.

```
// containers stdin
{"name":"event1"}
{"name":"event2"}
// containers stdout
{"status":200,"body":"Processed event1"}
{"status":200,"body":"Processed event2"}
```

So the communication layer will read events from `stdin` line by line and call lambda without shutting down the container. Then it will write the result of lambda execution as JSON string + `\n` to `stdout`, so it can be either seen by the developer or picked up by the wrapper (like SAM CLI).

This should also be pretty easy to integrate into existing [NPM module](https://github.com/lambci/docker-lambda#build-examples).

Some other questions:
- Should it be a new container type in addition to `run` and `build` or updated `run` container? Former sounds more reasonable to not break backwards compatibility, but brings some maintenance overhead.
- How to handle reporting? One option would be to wrap lambda result into another JSON object and add time/memory data as keys in the wrapper object. Is it even relevant here?

Limitations:
- Restarting the server, when code changes is out of scope. The primary use case for this feature is to be able to run tests faster and reloading is not very relevant there.

Does this sound reasonable? 
I plan to work on the proof of concept in the coming days.

**Comment by devoto13** [comment]
Thanks for you feedback, @mhart! I see how this protocol may be problematic in practice.

Another approach may be to expose HTTP API from the container, but it will involve much more effort and will make usage more complex comparing to `stdin`/`stdout` approach. I would prefer to not go this road.

What if we prefix handler result lines with some Unicode character to make them distinguishable from the lines, which were output by the handler itself? While it is impossible to guarantee that this character won't be printed by the users' code I think it is possible to pick the character, which will minimize the risk. We may also allow to customize this control character using an environment variable to allow users to solve possible collisions.

Control character may be either invisible (like [zero-width space](https://www.fileformat.info/info/unicode/char/200B/index.htm)) or meaningful (like [an arrow](http://www.fileformat.info/info/unicode/char/279d/index.htm)). In this case direct usage won't print something weird and wrappers will be able to use it.
